crates/macro/src/html_tree/html_component.rs

```diff
@@ -388,28 +392,94 @@ struct ListProps {
     props: Vec<HtmlProp>,
     node_ref: Option<Expr>,
 }
-
-impl Parse for ListProps {
-    fn parse(input: ParseStream) -> ParseResult<Self> {
+impl ListProps {
+    fn collect_props(input: ParseStream) -> ParseResult<Vec<HtmlProp>> {
         let mut props: Vec<HtmlProp> = Vec::new();
         while HtmlProp::peek(input.cursor()).is_some() {
             props.push(input.parse::<HtmlProp>()?);
         }
+        Ok(props)
+    }
 
+    fn remove_refs(mut props: Vec<HtmlProp>) -> ListProps {
         let ref_position = props.iter().position(|p| p.label.to_string() == "ref");
         let node_ref = ref_position.map(|i| props.remove(i).value);
-        for prop in &props {
+        ListProps { props, node_ref }
+    }
+
+    fn apply_edge_cases(props: &Vec<HtmlProp>) -> Result<(), syn::Error> {
```

Comment:
@jstarry What do you think about refactor of this [function](https://github.com/yewstack/yew/blob/master/crates/macro/src/html_tree/html_component.rs#L393) ?
It looks like I have to apply same [conditions](https://github.com/yewstack/yew/blob/master/crates/macro/src/html_tree/html_component.rs#L402) for both  [WithProps](https://github.com/yewstack/yew/blob/master/crates/macro/src/html_tree/html_component.rs#L439) and [ListProps](https://github.com/yewstack/yew/blob/master/crates/macro/src/html_tree/html_component.rs#L392) logic.
Then I have noticed that I have to do same work for handling `to many refs` error  in `with props` [logic](https://github.com/yewstack/yew/blob/master/crates/macro/src/html_tree/html_component.rs#L439).

So, I created a [HashMap](https://github.com/yewstack/yew/pull/960/files#diff-a8c81c5f22d5371fcf4d0ff922a224b4R411) with all edge cases. 
For example: if you want apply only `ref` edge case, you can pass "ref" as a second argument to `apply_edge_cases`. This let you use same logic for  both `ListProps` and `WithProps`.

If in the future, you would like to add more adge cases, you can only insert new `[key, valye]` to the hash map.
I just don't want to copy/paste logic from one function to another.

Reply:
@captain-yossarian nice! Here are my thoughts:

> It looks like I have to apply same conditions for both WithProps and ListProps logic.

Good point, how about we pull that common logic out of `WithProps` and `ListProps` and handle it `Props` instead? I'm imagining something like:

1. Parse list of props
2. Parse `with props`
3. Parse list of props
4. Check for double refs from 1. and 3.

> So, I created a HashMap with all edge cases

I'm not convinced this is really helping much, I think the for loop and if statements that we had before were clearer. We don't need to generalize this too much.

> I just don't want to copy/paste logic from one function to another.

Yeah, generally it's good to avoid copy/paste logic. I think in this case, we can avoid it by handling common paths at the `Props` level

> you can concatenate all errors in one string and output it to the user.

I think one error at a time is fine. Also, you can collect a list of results into one: https://doc.rust-lang.org/std/result/enum.Result.html#impl-FromIterator%3CResult%3CA%2C%20E%3E%3E

Reply:
With this approach, I think we can allow `ref` to come before `with props`

Reply:
@jstarry I have a problem with props type.
With `return_type` I can catch all kind of errors. It works perfectly.
But it works only with `let _ = Props::collect_props(input)?; ` 433 line
If I remove this line, app throw an error in this place: `return Err(input.error("MY"));` 464 line.

Do you have any thought why does it is happening?

Reply:
@captain-yossarian yeah the issue is that you are consuming the `ParseStream` in `get_type` and so there are no tokens left to parse. Look at how `Peek` implementations work. They use a cursor which can read without consuming any tokens from the `ParseStream`.

Reply:
@jstarry not quite. [Here](https://github.com/yewstack/yew/pull/960/files#diff-a8c81c5f22d5371fcf4d0ff922a224b4R382) I making a fork. If you log `input.cursor().token_stream().to_string()` at the end of `get_type` you will get all initial stream